vulkan_generator/vulkan_gen/__init__.py

```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2022 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This is the top level package for Vulkan Code Generator"""
```

Comment:
What's the difference between `vulkan_generator` and `vulkan_gen`? Just get rid of this package and use `vulkan_generator` as the top level package, since it's at the top level.

Reply:
Not even that, apparently using `--recursive` solves the problem

Reply:
For the build, we are not sure about bazel yet.

We need to run pytest, pylint and run the generator. There seems to be some workaround for pylint but maybe we should just use a shell script to trigger all of it as part of the build process different than bazel.

Reply:
And maybe we can trigger that shell from bazel if needed

Reply:
Also we need to add the pylint into presubmit, pytest into linux-test etc. but @purvisa-at-google-com  accepted to do that work in parallel so I think those can go to another PR.

Reply:
This looks like an interesting idea:

https://stackoverflow.com/a/67389420

I.e. Allow Bazel to manage your Python development dependencies, rather than relying on developers to install things (and who knows what versions of various Python tools/libraries they will get), and use a pytest test to run the linters, formatters, etc. so that they can easily be run just like a traditional pytest test (and so can be run via Bazel too).

If not Bazel, I would suggest at least using something from the Python ecosystem to manage the Python development dependencies, such as pipenv, as we used for gfauto:

https://pipenv.pypa.io/

https://github.com/google/graphicsfuzz/tree/master/gfauto#setup

Or [Poetry](https://python-poetry.org/docs/basic-usage/). Never used it, but is a commonly-suggested alternative to pipenv, and looks very nice.

You could still use the "pytest test to run the linters" approach regardless, I suppose.

Reply:
I think that making linter and/or test "targets" (in CMake or Bazel) is orthogonal to wanting to be able to make a change and build locally without being blocked by failing tests. AGI/GAPID used to have many go test targets that were failing, but they never blocked the build, nor were they run as part of the CI (instead, specific targets were were built and tested). But I don't have strong opinions about it. I think there are still two issues to consider, though:

- Managing your Python dependencies via Bazel, Pipenv, or Poetry. If you don't do this, Python tools/libraries (like pylint) will continuously update behind your back. A commit that passes presubmit today might fail presubmit in a week. A commit that passes on one machine today might fail on another machine today. Having said this, there is no harm is getting things started without proper dependency management, as it may not cause a lot of issues in practice if you are not using too many tools and pylint plugins. If you are already using Bazel to manage some of your build and linting dependencies, it perhaps makes sense to consider using Bazel to manage your Python dependencies. On the other hand, Bazel can feel quite incompatible with the language (i.e. Python) ecosystem to the point where it can be nice to avoid using it, in some cases.

- Adding script, lint, and test targets to Bazel, or adding certain scripts and linting procedures as pylint "tests", or just adding some standard (Bash? Python?) scripts that run code generation and linting (without needing any arguments). If you use Bazel to manage your Python dependencies, it can also make sense to ensure your Python scripts, tests, and linters can be run via Bazel targets. This unifies the way developers run scripts and tools (i.e. always via Bazel) and ensures dependencies are consistent. Also, for code generation, the dependency between the Python source files and generated C++ files can be accurately expressed. Alternatively, you might skip using Bazel and, instead, check in the generated code, ask developers to re-generate the C++ code via Python when needed, and enforce up-to-date generated code in the CI by running the code generation and checking that there were no changes (or, by ensuring the code generation scripts support a "check" or "dry run" option). Even if you don't use Bazel to run the scripts, there could still be a benefit to adding a special pytest "test" that actually just runs the linter, and another for running code generation, etc.: the benefit is that pytest tests have a standard way of being run, and so you no longer need to {"ask developers to run","document","manage bash scripts for"} a list of commands with various arguments in order to run the checks and/or code generation. E.g. `pylint --rcfile=vulkan_generator/pylintrc vulkan_generator/**/**.py`. Developers will almost certainly want to be able to run pylint locally sometimes, and if you encode all the arguments within a pytest "test" then the CI can run that test (along with the other tests), and so can developers (if they want). But actually, this is quite unusual, and there are some unusual consequences: the linter "test" would technically depend on "all Python source files", and the code generation "test" would technically produce output files / modify C++ source files. In gfauto, we just had bash scripts that had to be run from a specific directory, with no arguments, which can also work.